proposals/p2760.md

```diff
@@ -0,0 +1,794 @@
+  fn Print[self: Self]();
+}
+```
+
+The way this works is that `Vector.Print` is equivalent to
+`Vector.(Printable.Print)`, which may or may not be defined. The name
+`Vector.Print` can no longer be conditional, and the meaning of that name is
+fixed. However, the implementation of `Printable` for `Vector(T)` may not exist
+for some types `T`.
+
+### Adapters
+
+What was previously written:
+
+```
+class C;
+// Forward declarations of adapters.
+adapter A for C;
+adapter E extends C;
+
+// Definitions of an adapter.
+adapter A for C {
+  ...
+}
+// Definition of an extending adapter.
+adapter E extends C {
+  ...
+}
+```
+
+is now written:
+
+```
+class C;
+// Forward declarations of adapters.
+class A;
+class E;
+
+// Definitions of an adapter.
+class A {
+  adapt C;
+  ...
+}
+// Definition of an extending adapter.
+class E {
+  extend adapt C;
+  ...
+}
+```
+
+Note:
+
+-   Adapters are now a special case of classes, not a distinct top-level
+    declaration.
+-   Classes with `adapt` still must not contain anything that was previously
+    forbidden for adapters: no fields, no base class, no virtual methods, no
+    implementations of virtual methods, and so on.
+-   The `adapt` declaration must appear before
+    [mixin declarations](#future-work-mixins), if any.
+-   The syntax for an `adapt` declaration inside a class body is:
+
+    > [`extend`] `adapt` _type-expression_ `;`
+
+### Interfaces
+
+What was previously written:
+
+```
+interface A { let T:! Type; }
+interface B { let U:! Type; }
+interface C { let V:! Type; }
+interface D { let W:! Type; }
+
+interface I {
+  // Interface is incorporated into `I`.
+  extends A where .T == i32;
+  extends B where .U == i32;
+
+  // No impact on `I`s interface, but an implementation must exist.
+  impl as C where .V == i32;
+  impl as D where .W == i32;
+}
+```
+
+is now written:
+
+```
+interface A { let T:! Type; }
+interface B { let U:! Type; }
+interface C { let V:! Type; }
+interface D { let W:! Type; }
+
+interface I {
+  // `A`'s interface is incorporated into `I`.
+  extend A;
+
+  // Anything that could be in a `where` clause can
+  // be in a `require` declaration:
+  require A.T == i32;
+
+  // Can combine `extend` with `where`.
+  extend B where B.U == i32;
+
+  // No impact on `I`s interface, but an
+  // implementation must exist.
+  require Self impls C where .V == i32;
+
+  // Again the `where` clause can be located
+  // in a separate `require` declaration.
+  require Self impls D;
+  require D.W == i32;
+}
+```
+
+Notes:
+
+-   The same change applies to named constraints, and is intended to be used in
+    the future for named predicates used as template constraints.
+-   One syntax for constraints in either `where` clauses or `require`
+    declarations.
+-   Want to open up the syntax to expressing more general constraints.
+-   Syntax for a `require` declaration in an interface or named constraint:
+
+    > `require` _where-constraint-argument_ `;`
```

Comment:
I remember there being a question about whether rewrite constraints are disallowed here, or whether they're allowed but don't result in rewrites being performed. (Might have been in the context of #2409.) But I don't remember what the conclusion was.

Reply:
I would welcome restrictions here -- I would prefer that there were fewer ways of writing equivalent interfaces. In fact, I'm pretty unhappy with the freedom shown in the example above. I think the only requirement is that the syntax matches across `where` and `require` for constraints that are allowed in both.

Reply:
Agreed with @josh11b here, I'd prefer to allow the minimal set of things, as long as we don't have *divergence* between `where` and `requires`.

Reply:
Addressed in https://github.com/carbon-language/carbon-lang/pull/2760/commits/e8f9974245326c1efe4b164650bd6e9477513c13